rts/lib/lua/src/lbaselib.cpp

```diff
@@ -236,10 +236,15 @@ static int luaB_next (lua_State *L) {
 
 
 static int luaB_pairs (lua_State *L) {
-  luaL_checktype(L, 1, LUA_TTABLE);
-  lua_pushvalue(L, lua_upvalueindex(1));  /* return generator, */
-  lua_pushvalue(L, 1);  /* state, */
-  lua_pushnil(L);  /* and initial value */
+  luaL_checkany(L, 1);
+  if (luaL_getmetafield(L, 1, "__pairs") == LUA_TNIL) {  /* no metamethod? */
+    lua_pushcfunction(L, luaB_next);  /* return generator, */
```

Comment:
Why the change from pushing `lua_upvalueindex(1)` to pushing `luaB_next` directly? Can the generator ever not be `next`? Maybe there are lifetime differences (won't this make a new instance of the C function each time)? What about other functions that use the "push upvalue 1" pattern?

Reply:
the code I have there is (almost) exactly how the latest version of Lua does it.

https://github.com/lua/lua/blob/c1dc08e8e8e22af9902a6341b4a9a9a7811954cc/lbaselib.c#L284-L296

```cpp
static int luaB_pairs (lua_State *L) {
  luaL_checkany(L, 1);
  if (luaL_getmetafield(L, 1, "__pairs") == LUA_TNIL) {  /* no metamethod? */
    lua_pushcfunction(L, luaB_next);  /* will return generator, */
    lua_pushvalue(L, 1);  /* state, */
    lua_pushnil(L);  /* and initial value */
  }
  else {
    lua_pushvalue(L, 1);  /* argument 'self' to metamethod */
    lua_callk(L, 1, 3, 0, pairscont);  /* get 3 values from metamethod */
  }
  return 3;
}
```

Reply:
They've also completely dropped the use of "auxopen" in setting up the base functions.

Which was a convenience method that make a closure so that the other function needed was there was always there as the first upvalue.

```cpp
static void auxopen (lua_State *L, const char *name,
                     lua_CFunction f, lua_CFunction u) {
  lua_pushcfunction(L, u);
  lua_pushcclosure(L, f, 1);
  lua_setfield(L, -2, name);
}
```

I'll also look and see if there was also any changes to how lua_pushcfunction worked in the newer versions of Lua

Reply:
...
seems that there was a change that added "light C functions", that allowed the use of upvalues to be removed
the idea being that in certain cases C functions could be used without needing to make a full closure

the newest version of pushcclosure is like so- which avoids GC overhead
```cpp
LUA_API void lua_pushcclosure (lua_State *L, lua_CFunction fn, int n) {
  lua_lock(L);
  if (n == 0) {
    setfvalue(s2v(L->top.p), fn);
    api_incr_top(L);
  }
  else {
    int i;
    CClosure *cl;
    api_checkpop(L, n);
    api_check(L, n <= MAXUPVAL, "upvalue index too large");
    cl = luaF_newCclosure(L, n);
    cl->f = fn;
    for (i = 0; i < n; i++) {
      setobj2n(L, &cl->upvalue[i], s2v(L->top.p - n + i));
      /* does not need barrier because closure is white */
      lua_assert(iswhite(cl));
    }
    L->top.p -= n;
    setclCvalue(L, s2v(L->top.p), cl);
    api_incr_top(L);
    luaC_checkGC(L);
  }
  lua_unlock(L);
}
```

this is quite the rabbit hole to potentially go down to backport all this supporting code as well- thanks for bringing up your `won't this make a new instance of the C function each time?` concern.

I'll change the code to use the upvalue since "light C functions" do not seem to be supported in our jank version of Lua

Reply:
It seems that it was realized that c functions... don't need a fully closure in most cases.